sites/cheerpx/src/content/docs/11-guides/input-output.md

```diff
@@ -0,0 +1,117 @@
+---
+title: Input and output
+description: Techniques for getting data in and out of the CheerpX virtual machine
+---
+
+## CheerpX console input and output
+
+CheerpX provides two options for handling console input and output: a _built-in_ console and a _custom_ console. The purpose of the console is to interact with the running application via standard input and output.
+
+### Built-in console
+
+CheerpX comes with a _built-in_ console that mimics traditional console behavior. The _built-in_ console supports standard input and output over a HTML `<pre>` (pre-formatted) element. Keyboard input works provided the element is in focus.
+
+For more information, visit the [CheerpX console] reference.
+
+### Custom console
+
+CheerpX also supports a _custom_ console that allows developers to programmatically capture output and provide input. This feature is useful to integrate with a more complete terminal implementation such as [xterm.js]. We use xterm.js ourselves for our public WebVM environment.
+
+Another possible use for the custom console is accumulating program output into a JavaScript string. You can achieve this with the following snippet:
+
+```js
+const encoder = new TextEncoder("utf-8");
+
+let accumulatedOutput = ""; // Initialize an empty string to accumulate output
+
+const send = cx.setCustomConsole((buf) => {
+	const string = new TextDecoder("utf-8").decode(buf);
+	accumulatedOutput += string; // Accumulate the output
+	console.log(accumulatedOutput); // log the accumulated output
+});
+
+// Send a string
+const str = "Hello, custom console!\n";
+const encodedStr = encoder.encode(str);
+
+for (let i = 0; i < encodedStr.length; i++) {
+	send(encodedStr[i]);
+}
+```
+
+For more details on customizing the console, see [CheerpX Custom console].
+
+## Reading Files Using IDBDevice and Redirecting Output
+
+`IDBDevice` provides a persistent, read-write filesystem using the browser’s IndexedDB. It’s ideal for storing data that should persist between sessions. You can use the `readFileAsBlob` method to read files from an `IDBDevice` as Blob objects.
+
+To make a file accessible, you can copy files using commands within the virtual machine. Here’s an example on how to do it:
+
+```js
+const filesDevice = await CheerpX.IDBDevice.create("files");
+const cx = await CheerpX.Linux.create({
+	// Mount the IDBDevice
+	mounts: [
+		{ type: "ext2", path: "/", dev: overlayDevice },
+		{ type: "dir", path: "/files", dev: filesDevice },
+	],
+});
+```
+
+Now, you can copy files as follows:
+
+```js
+await cx.run("/bin/cp", ["/source_file", "/files/destination_file"], {
+	env: [
```

Comment:
Neither env, nor cwd are required. You can completely remove the optional arguments.